documentation/high-level-introduction.md

```diff
@@ -0,0 +1,257 @@
+    - Property and Item Group will set items in the global context
+    - UsingTask will load task assemblies and prepare to invoke them
+    - Targets will define then and their ordering\inputs\output
+      - **Their contents will not be executed yet**
+  - This will respect conditions, and be done top to bottom based upon include order
+- Target Eval
+  - Starting with the defined “DefaultTargets", which for our purposes is always “Build” determine a list of target to invoke and their ordering
+  - Invoke one at a time, comparing their Input\Output against file timestamps like make
+
+
+## "SDK style” project and nuget injection
+  - SDK style projects and nugets are beyond the scope of this document 
+  - One important thing to note is the “magic” injection of files
+    - When referenced the props file will be “auto injected” to the top of your project and the targets injected at the very bottom. 
+    - Thus you can include files for these project types you don’t explicitly see referenced in the csproj
+
+
+## Debugging
+  - Easily the most important thing is to get a full diagnostic build. This will list all variables set and the ordering\execution of each target.
+  - You can add debug “Console.WriteLine”s by using the built in MessageTask:
+
+
+            <Message Text="Project File Name = $(MSBuildProjectFile)" />
+
+    - **Do note these must be inside targets**. They are only executed during Target Eval time.
+    - If you put them inside ItemGroups or PropertyGroups they will be ignored.
+  - There is a new “binary log” hotness which acts as a full diagnostics build but is faster and contains even more information. You can enable it by passing `/bl`  to msbuild. Read more about it [here](https://github.com/Microsoft/msbuild/blob/master/documentation/wiki/Binary-Log.md). 
+    - The support for loading .binlog files in Visual Studio for Mac is still somewhat new
+  - If you are developing on Windows you can add a `System.Diagnostics.Debugger.Launch()` invocation and attach Visual Studio to the process.
+
+
+## Gotchas
+  - Visual Studio for Mac will cache your task assemblies but not your target\prop files, usually.
+    - This means if you make changes to a task assembly, your changes may not be reflected until you restart your IDE.
+    - This is why I always test using msbuild from the command line OR write an nunit test that invokes msbuild directly.
+  - Ordering matters, dearly
+    - If file A defines and variable and B uses it, then A must be included before B in all cases.
+    - If you get this wrong, then B will start getting empty string and not the valid variable
+    - This is why people generally use props and targets files to keep the order right
+    - You can not safely refactor msbuild without serious testing. 
+  - Everything is a string
+    - Typos will ruin your day, since you’ll get empty strings in places you do not expect
+    - Make sure your quotes are correct else you will be very confused
```

Comment:
...quotes are correct, or you will be very confused.

It would be good to be more consistent about whether sentences in bullet points should end in "." or not.